### PR TITLE
elf2efi: add one more section that should be ignored

### LINKED ISSUE #28520: Build failure with DT_RELR (-Wl,-z,pack-relative-relocs)
### systemd version the issue has been seen with

sys-apps/systemd-254_rc2

### Used distribution

Gentoo

### Linux kernel version used

6.4.6

### CPU architectures issue was seen on

x86_64

### Component

systemd-boot

### Expected behaviour you didn't see

Successful build.

### Unexpected behaviour you saw

Build failure when building `linuxx64.efi.stub` with `RuntimeError: Unknown section .relr.dyn, refusing.`.

### Steps to reproduce the problem

Build systemd with `LDFLAGS="-Wl,-z,pack-relative-relocs"` to use `DT_RELR`.

### Additional program output to the terminal or log subsystem illustrating the issue

Full build log: [build.log.gz](https://github.com/systemd/systemd/files/12167684/build.log.gz)

```sh
[661/2632] /var/tmp/portage/sys-apps/systemd-254_rc3/work/systemd-254-rc3/tools/elf2efi.py --version-major=254 --version-minor=0 --efi-major=1 --efi-minor=1 --subsystem=10 --minimum-sections=0 src/boot/efi/li
nuxx64.elf.stub src/boot/efi/linuxx64.efi.stub
FAILED: src/boot/efi/linuxx64.efi.stub
/var/tmp/portage/sys-apps/systemd-254_rc3/work/systemd-254-rc3/tools/elf2efi.py --version-major=254 --version-minor=0 --efi-major=1 --efi-minor=1 --subsystem=10 --minimum-sections=0 src/boot/efi/linuxx64.elf.
stub src/boot/efi/linuxx64.efi.stub
Traceback (most recent call last):
  File "/var/tmp/portage/sys-apps/systemd-254_rc3/work/systemd-254-rc3/tools/elf2efi.py", line 593, in <module>
    main()
  File "/var/tmp/portage/sys-apps/systemd-254_rc3/work/systemd-254-rc3/tools/elf2efi.py", line 589, in main
    elf2efi(parser.parse_args())
  File "/var/tmp/portage/sys-apps/systemd-254_rc3/work/systemd-254-rc3/tools/elf2efi.py", line 492, in elf2efi
    sections = copy_sections(elf, opt)
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/tmp/portage/sys-apps/systemd-254_rc3/work/systemd-254-rc3/tools/elf2efi.py", line 288, in copy_sections
    if not use_section(elf_s):
           ^^^^^^^^^^^^^^^^^^
  File "/var/tmp/portage/sys-apps/systemd-254_rc3/work/systemd-254-rc3/tools/elf2efi.py", line 256, in use_section
    raise RuntimeError(f"Unknown section {elf_s.name}, refusing.")
RuntimeError: Unknown section .relr.dyn, refusing.
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #28520.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: tools/elf2efi.py [hunk 1]
```diff
@@ -231,6 +231,7 @@ def use_section(elf_s: ELFSection) -> bool:
         ".note.gnu.build-id",
         ".rel.dyn",
         ".rela.dyn",
+        ".relr.dyn",
```
Comment (thesamesam) [inline tools/elf2efi.py:234]: Heh, I should've actually looked at the file, and then I could've given a PR myself. Thanks!

Comment (bluca) [reply tools/elf2efi.py:234]: can you confirm this fixes the issue for you?

**Comment by medhefgo** [comment]
This is a bad idea because elf2efi does not handle `DT_RELR` relocations. The `ld` manpage says it shouldn't generate any in the first place when creating a pie, so I wonder why it's created. Anyway, I think we should pass `-z nopack-relative-relocs` to efi ld flags.

File: tools/elf2efi.py (hunk 1 above)
Comment (thesamesam) [reply tools/elf2efi.py:234]: Yeah, it fixes it for me.

**Comment by bluca** [comment]
> This is a bad idea because elf2efi does not handle `DT_RELR` relocations. The `ld` manpage says it shouldn't generate any in the first place when creating a pie, so I wonder why it's created. Anyway, I think we should pass `-z nopack-relative-relocs` to efi ld flags.
> 
> (Alternatively, one could also just teach it how to handle `DT_RELR`.)

Do you have time to send a PR for this? Would be nice to fix for 254

**Comment by medhefgo** [comment]
> Do you have time to send a PR for this? Would be nice to fix for 254

I can do it later. But all you need is to add the flag here: https://github.com/systemd/systemd/blob/859ecf0aaa51ac1f5bcefab3fb1ba6156de9d851/src/boot/efi/meson.build#L173

**Comment by bluca** [comment]
@yuwata are you still around to update this as suggested?

**Comment by bluca** [comment]
--> https://github.com/systemd/systemd/pull/28523

**Pull request closed without merging**